From 92d3454475dc1413517ff60b999b14d0a308e994 Mon Sep 17 00:00:00 2001
From: Peter Hawkins <phawkins@google.com>
Date: Mon, 16 Oct 2023 16:01:34 -0700
Subject: [PATCH] [XLA:GPU] Fix bug in all-to-all for complex data types.

The multiplier for complex data types wasn't being applied correctly; the chunk_bytes calculation double-applied the multiplier.

Fixes https://github.com/google/jax/issues/18122

PiperOrigin-RevId: 573955671
---
 xla/python/xla_client.py                 |  2 +-
 xla/service/gpu/nccl_all_to_all_thunk.cc | 22 ++++++++++------------
 2 files changed, 11 insertions(+), 13 deletions(-)

diff --git a/xla/python/xla_client.py b/xla/python/xla_client.py
index a07596730a3d3..3d61b9a6188f6 100644
--- a/xla/python/xla_client.py
+++ b/xla/python/xla_client.py
@@ -47,7 +47,7 @@
 
 # Just an internal arbitrary increasing number to help with backward-compatible
 # changes. In JAX, reference this via jax._src.lib.xla_extension_version.
-_version = 206
+_version = 207
 
 # Version number for MLIR:Python components.
 mlir_api_version = 54
diff --git a/xla/service/gpu/nccl_all_to_all_thunk.cc b/xla/service/gpu/nccl_all_to_all_thunk.cc
index 2dcf4ac8fe79f..a94451fb62822 100644
--- a/xla/service/gpu/nccl_all_to_all_thunk.cc
+++ b/xla/service/gpu/nccl_all_to_all_thunk.cc
@@ -128,9 +128,8 @@ Status RunAllToAll(bool has_split_dimension,
       TF_ASSIGN_OR_RETURN(auto dtype_and_multiplier,
                           ToNcclDataTypeAndCountMultiplier(
                               buffer.element_type, Thunk::kNcclAllToAll));
-      ncclDataType_t dtype = dtype_and_multiplier.first;
-      int64_t element_count =
-          buffer.element_count * dtype_and_multiplier.second;
+      auto [dtype, multiplier] = dtype_and_multiplier;
+      int64_t element_count = buffer.element_count;
 
       TF_RET_CHECK(element_count % num_participants == 0)
           << "Buffer was not an exact multiple of the number of participants.";
@@ -142,21 +141,21 @@ Status RunAllToAll(bool has_split_dimension,
         VLOG(3) << absl::StreamFormat(
             "Calling ncclSend(sendbuff=%p, count=%d, peer=%d "
             "comm=%p, stream=%p)",
-            send_buffer + rank * chunk_bytes, chunk_elements, rank,
+            send_buffer + rank * chunk_bytes, chunk_elements * multiplier, rank,
             static_cast<const void*>(comm), gpu_stream);
         XLA_CUDA_RETURN_IF_ERROR(ncclSend(send_buffer + rank * chunk_bytes,
-                                          chunk_elements, dtype, rank, comm,
-                                          gpu_stream));
+                                          chunk_elements * multiplier, dtype,
+                                          rank, comm, gpu_stream));
 
         VLOG(3) << absl::StreamFormat(
             "Calling ncclRecv(recvbuff=%p, count=%d, peer=%d "
             "comm=%p, stream=%p)",
-            recv_buffer + rank * chunk_bytes, chunk_elements, rank,
+            recv_buffer + rank * chunk_bytes, chunk_elements * multiplier, rank,
             static_cast<const void*>(comm), gpu_stream);
 
         XLA_CUDA_RETURN_IF_ERROR(ncclRecv(recv_buffer + rank * chunk_bytes,
-                                          chunk_elements, dtype, rank, comm,
-                                          gpu_stream));
+                                          chunk_elements * multiplier, dtype,
+                                          rank, comm, gpu_stream));
       }
     }
   } else {
@@ -173,9 +172,8 @@ Status RunAllToAll(bool has_split_dimension,
       TF_ASSIGN_OR_RETURN(auto dtype_and_multiplier,
                           ToNcclDataTypeAndCountMultiplier(
                               buffer.element_type, Thunk::kNcclAllToAll));
-      ncclDataType_t dtype = dtype_and_multiplier.first;
-      int64_t element_count =
-          buffer.element_count * dtype_and_multiplier.second;
+      auto [dtype, multiplier] = dtype_and_multiplier;
+      int64_t element_count = buffer.element_count * multiplier;
 
       VLOG(3) << absl::StreamFormat(
           "Calling ncclSend(sendbuff=%p, count=%d, peer=%d "
